packages/ts/models/src/builders.ts

```diff
@@ -0,0 +1,254 @@
+ * @typeParam CV - The current value type of the model. It changes as the model
+ * is being built and defines if the
+ * {@link ObjectModelBuilder.build} method can be called.
+ * @typeParam EX - The extra properties of the model.
+ * @typeParam F - The flags for the model constructor that allow to determine
+ * specific characteristics of the model.
+ *
+ * @internal
+ */
+export class ObjectModelBuilder<
+  V extends AnyObject,
+  CV extends AnyObject = EmptyObject,
+  EX extends AnyObject = EmptyObject,
+  F extends Flags = { named: false; selfRefKeys: never },
+> extends CoreModelBuilder<V, EX, F> {
+  constructor(base: Model) {
+    super(base, (m) => {
+      const result = create(null);
+
+      // eslint-disable-next-line no-restricted-syntax
+      for (const key in m) {
+        defineProperty(result, key, {
+          enumerable: true,
+          get: () => (m[key as keyof Model<V, EX>] as Model)[$defaultValue],
+        });
+      }
+
+      return result;
+    });
+  }
+
+  /**
+   * The method that should follow the {@link m.extend} method. It allows to
+   * declare the extension for the model and properly name it.
+   *
+   * @param name - The name of the model.
+   */
+  object<NV extends AnyObject>(
+    this: F['named'] extends false ? this : never,
+    name: string,
+  ): ObjectModelBuilder<NV & V, CV, EX, { named: true; selfRefKeys: F['selfRefKeys'] }> {
+    return this.name(name) as any;
+  }
+
+  /**
+   * {@inheritDoc CoreModelBuilder.define}
+   */
+  declare ['define']: <DK extends symbol, DV>(
+    key: DK,
+    value: TypedPropertyDescriptor<DV>,
+  ) => ObjectModelBuilder<V, CV, EX & Readonly<Record<DK, DV>>, F>;
+
+  /**
+   * {@inheritDoc CoreModelBuilder.meta}
+   */
+  declare ['meta']: (value: ModelMetadata) => this;
+
+  /**
+   * Defines a new model property on the model. Unlike the
+   * {@link ObjectModelBuilder.define}, this property is public and allows the
+   * user to interact with the model data structure. It also updates the current
+   * value type of the model to make it closer to the final value type.
+   *
+   * @param key - The key of the property.
+   * @param model - The model of the property value. You can also provide a
+   * function that produces the model based on the current model.
+   * @param options - Additional options for the property.
+   */
+  property<PK extends string & keyof V, EXK extends AnyObject = EmptyObject>(
+    key: PK,
+    model: Model<V[PK], EXK> | ((model: Model<V, EX & Readonly<Record<PK, Model<V[PK], EXK>>>>) => Model<V[PK], EXK>),
+    options?: ModelBuilderPropertyOptions,
+  ): Extract<V[PK], V> extends never
+    ? ObjectModelBuilder<V, CV & Readonly<Record<PK, V[PK]>>, EX & Readonly<Record<PK, Model<V[PK], EXK>>>, F>
+    : ObjectModelBuilder<
+        V,
+        CV & Readonly<Record<PK, V[PK]>>,
+        EX,
+        { named: F['named']; selfRefKeys: F['selfRefKeys'] | PK }
+      > {
+    defineProperty(this[$model], key, {
```

Comment:
The signature of this method is so complex that it requires a deeper explanation.